tests: k8s: failed exec from k8s-pid-ns.bats

Example:

https://github.com/kata-containers/kata-containers/actions/runs/7999467748/job/21869181017?pr=9128

The pod state looks Ready to me, yet kubectl exec fails with: "error: unable to upgrade connection: container not found ("first-test-container")"

Discussions on the Internet suggest that sometimes that happens on slower pods and that the exec/connection to the pod should be retried.

```
[run_kubernetes_tests.sh:142] INFO: Executing k8s-pid-ns.bats
1..1
not ok 1 Check PID namespaces
# (in test file k8s-pid-ns.bats, line 37)
#   `-- $ps_command | grep "/pause")' failed
# [bats-exec-test:55] INFO: k8s configured to use runtimeclass
# [bats-exec-test:191] INFO: /home/runner/work/kata-containers/kata-containers/tests/integration/kubernetes/runtimeclass_workloads_work/genpolicy.zU5hExx2Mi/genpolicy-settings.json: allowing exec: ps
# [bats-exec-test:210] INFO: /home/runner/work/kata-containers/kata-containers/tests/integration/kubernetes/runtimeclass_workloads_work/genpolicy.zU5hExx2Mi/genpolicy-settings.json: allowing ReadStreamRequest
# [bats-exec-test:178] INFO: Executing: RUST_LOG=info /opt/kata/bin/genpolicy -u -y /home/runner/work/kata-containers/kata-containers/tests/integration/kubernetes/runtimeclass_workloads_work/busybox-pod.yaml -p /home/runner/work/kata-containers/kata-containers/tests/integration/kubernetes/runtimeclass_workloads_work/genpolicy.zU5hExx2Mi/rules.rego -j /home/runner/work/kata-containers/kata-containers/tests/integration/kubernetes/runtimeclass_workloads_work/genpolicy.zU5hExx2Mi/genpolicy-settings.json
# [2024-02-22T15:49:40Z INFO  genpolicy::registry] ============================================
# [2024-02-22T15:49:40Z INFO  genpolicy::registry] Pulling manifest and config for "quay.io/prometheus/busybox:latest"
# [2024-02-22T15:49:41Z INFO  genpolicy::registry] Using cached file "layers_cache/sha256-9fa9226be034e47923c0457d916aa68474cdfb23af8d4525e9baeebc4760977a.verity"
# [2024-02-22T15:49:41Z INFO  genpolicy::registry] dm-verity root hash: 2e52d86498030b4b99318650826d3f121bdfac5fe7bbbea9d18c546d48003aa0
# [2024-02-22T15:49:41Z INFO  genpolicy::registry] Using cached file "layers_cache/sha256-1617e25568b2231fdd0d5caff63b06f6f7738d8d961f031c80e47d35aaec9733.verity"
# [2024-02-22T15:49:41Z INFO  genpolicy::registry] dm-verity root hash: 75ed5e78d30374b9557d32c17ae6d2e1b4afdc5efd128812e6c99a0f60708dba
# [2024-02-22T15:49:41Z INFO  genpolicy::registry] ============================================
# [2024-02-22T15:49:41Z INFO  genpolicy::registry] Pulling manifest and config for "quay.io/prometheus/busybox:latest"
# [2024-02-22T15:49:41Z INFO  genpolicy::registry] Using cached file "layers_cache/sha256-9fa9226be034e47923c0457d916aa68474cdfb23af8d4525e9baeebc4760977a.verity"
# [2024-02-22T15:49:41Z INFO  genpolicy::registry] dm-verity root hash: 2e52d86498030b4b99318650826d3f121bdfac5fe7bbbea9d18c546d48003aa0
# [2024-02-22T15:49:41Z INFO  genpolicy::registry] Using cached file "layers_cache/sha256-1617e25568b2231fdd0d5caff63b06f6f7738d8d961f031c80e47d35aaec9733.verity"
# [2024-02-22T15:49:41Z INFO  genpolicy::registry] dm-verity root hash: 75ed5e78d30374b9557d32c17ae6d2e1b4afdc5efd128812e6c99a0f60708dba
# [2024-02-22T15:49:41Z INFO  genpolicy::registry] ============================================
# [2024-02-22T15:49:41Z INFO  genpolicy::registry] Pulling manifest and config for "mcr.microsoft.com/oss/kubernetes/pause:3.6"
# [2024-02-22T15:49:41Z INFO  genpolicy::registry] Using cached file "layers_cache/sha256-5720cd9c19ca69b58202945924c37c9bd7b287ce1a88882098fd59a4292e7cd9.verity"
# [2024-02-22T15:49:41Z INFO  genpolicy::registry] dm-verity root hash: 817250f1a3e336da76f5bd3fa784e1b26d959b9c131876815ba2604048b70c18
# [2024-02-22T15:49:41Z INFO  genpolicy] Success!
# pod/busybox created
# pod/busybox condition met
# error: unable to upgrade connection: container not found ("first-test-container")
# Name:                busybox
# Namespace:           kata-containers-k8s-tests
# Priority:            0
# Runtime Class Name:  kata
# Service Account:     default
# Node:                aks-nodepool1-17044399-vmss000000/10.224.0.4
# Start Time:          Thu, 22 Feb 2024 15:49:42 +0000
# Labels:              <none>
# Annotations:         io.katacontainers.config.agent.policy:
#                        IyBDb3B5cmlnaHQgKGMpIDIwMjMgTWljcm9zb2Z0IENvcnBvcmF0aW9uCiMKIyBTUERYLUxpY2Vuc2UtSWRlbnRpZmllcjogQXBhY2hlLTIuMAojCnBhY2thZ2UgYWdlbnRfcG9saW...
#                      io.katacontainers.config.hypervisor.initrd: /opt/kata/share/kata-containers/kata-containers-initrd-mariner.img
#                      io.katacontainers.config.hypervisor.kernel: /usr/share/cloud-hypervisor/vmlinux.bin
# Status:              Running
# IP:                  10.244.0.46
# IPs:
#   IP:  10.244.0.46
# Containers:
#   first-test-container:
#     Container ID:  containerd://08216d9ae710840683ead0591d0bf39ef5a777810e0d2e2a2d39ee58d99a883b
#     Image:         quay.io/prometheus/busybox:latest
#     Image ID:      quay.io/prometheus/busybox@sha256:dfa54ef35e438b9e71ac5549159074576b6382f95ce1a434088e05fd6b730bc4
#     Port:          <none>
#     Host Port:     <none>
#     Command:
#       sleep
#       30
#     State:          Running
#       Started:      Thu, 22 Feb 2024 15:50:28 +0000
#     Ready:          True
#     Restart Count:  0
#     Environment:
#       CONTAINER_NAME:  first-test-container
#     Mounts:
#       /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-h8lx7 (ro)
#   second-test-container:
#     Container ID:  containerd://99c08dc34f0d10b76ea79fd8ec90186d656621bafddac937f92b0e3662a26da6
#     Image:         quay.io/prometheus/busybox:latest
#     Image ID:      quay.io/prometheus/busybox@sha256:dfa54ef35e438b9e71ac5549159074576b6382f95ce1a434088e05fd6b730bc4
#     Port:          <none>
#     Host Port:     <none>
#     Command:
#       sleep
#       30
#     State:          Running
#       Started:      Thu, 22 Feb 2024 15:51:01 +0000
#     Ready:          True
#     Restart Count:  0
#     Environment:
#       CONTAINER_NAME:  second-test-container
#     Mounts:
#       /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-h8lx7 (ro)
# Conditions:
#   Type              Status
#   Initialized       True
#   Ready             True
#   ContainersReady   True
#   PodScheduled      True
# Volumes:
#   kube-api-access-h8lx7:
#     Type:                    Projected (a volume that contains injected data from multiple sources)
#     TokenExpirationSeconds:  3607
#     ConfigMapName:           kube-root-ca.crt
#     ConfigMapOptional:       <nil>
#     DownwardAPI:             true
# QoS Class:                   BestEffort
# Node-Selectors:              katacontainers.io/kata-runtime=true
# Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
#                              node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
# Events:
#   Type    Reason     Age   From               Message
#   ----    ------     ----  ----               -------
#   Normal  Scheduled  79s   default-scheduler  Successfully assigned kata-containers-k8s-tests/busybox to aks-nodepool1-17044399-vmss000000
#   Normal  Pulling    65s   kubelet            Pulling image "quay.io/prometheus/busybox:latest"
#   Normal  Pulled     65s   kubelet            Successfully pulled image "quay.io/prometheus/busybox:latest" in 185.557043ms (185.564943ms including waiting)
#   Normal  Created    65s   kubelet            Created container first-test-container
#   Normal  Started    33s   kubelet            Started container first-test-container
#   Normal  Pulling    33s   kubelet            Pulling image "quay.io/prometheus/busybox:latest"
#   Normal  Pulled     33s   kubelet            Successfully pulled image "quay.io/prometheus/busybox:latest" in 131.516177ms (131.522377ms including waiting)
#   Normal  Created    33s   kubelet            Created container second-test-container
#   Normal  Started    0s    kubelet            Started container second-test-container
# pod "busybox" deleted
# [bats-exec-test:157] INFO: Deleting /home/runner/work/kata-containers/kata-containers/tests/integration/kubernetes/runtimeclass_workloads_work/genpolicy.zU5hExx2Mi
Error: Process completed with exit code 1.
```

--- a/tests/integration/kubernetes/k8s-exec.bats
+++ b/tests/integration/kubernetes/k8s-exec.bats
@@ -69,11 +69,11 @@ EOF"
 
 	## Cases for target container
 	### First container
-	container_name=$(kubectl exec $pod_name -c $first_container_name -- $env_command | grep CONTAINER_NAME)
+	container_name=$(kubectl exec $pod_name -c $first_container_name -- $env_command | grep CONTAINER_NAME | tr -d '\r')
 	[ "$container_name" == "CONTAINER_NAME=$first_container_name" ]
 
 	### Second container
-	container_name=$(kubectl exec $pod_name -c $second_container_name -- $env_command | grep CONTAINER_NAME)
+	container_name=$(kubectl exec $pod_name -c $second_container_name -- $env_command | grep CONTAINER_NAME | tr -d '\r')
 	[ "$container_name" == "CONTAINER_NAME=$second_container_name" ]
 
 }

--- a/tests/integration/kubernetes/k8s-pid-ns.bats
+++ b/tests/integration/kubernetes/k8s-pid-ns.bats
@@ -35,15 +35,16 @@ setup() {
 	kubectl wait --for=condition=Ready --timeout=$timeout pod $pod_name
 
 	# Check PID from first container
+	# Strip \r — containers with tty: true return \r\n line endings
 	first_pid_container=$(kubectl exec $pod_name -c $first_container_name \
-		-- $ps_command | grep "/pause")
+		-- $ps_command | grep "/pause" | tr -d '\r')
 	# Verify that is not empty
 	check_first_pid=$(echo $first_pid_container | wc -l)
 	[ "$check_first_pid" == "1" ]
 
 	# Check PID from second container
 	second_pid_container=$(kubectl exec $pod_name -c $second_container_name \
-		-- $ps_command | grep "/pause")
+		-- $ps_command | grep "/pause" | tr -d '\r')
 	# Verify that is not empty
 	check_second_pid=$(echo $second_pid_container | wc -l)
 	[ "$check_second_pid" == "1" ]

